aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -253,6 +251,259 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
+  }
+
+  Tensor result;
+
+  if (out.defined()) {
+    TORCH_CHECK(
+        dtype == out.dtype(),
+        "multi_dot(): expected out tensor to have dtype ",
+        dtype,
+        " but got ",
+        out.dtype());
+    TORCH_CHECK(
+        device == out.device(),
+        "multi_dot(): expected out tensor to be on device ",
+        device,
+        " but got ",
+        out.device());
+
+    at::native::resize_output(out, out_shape);
+
+    // View output as 2D for computations
+    result = out.view({tensors[0].size(0), tensors.back().size(-1)});
+  }
+
+  // The resize_ and view calls below are to ensure the
+  // output shape respects the original dimensionality of
+  // the first and last tensors which we are now viewed as 2D
+
+  if (has_empty_tensor) {
+    return out.defined() ? result.zero_()
+                         : at::zeros(out_shape, tensors[0].options());
+  }
+
+  if (tensors.size() == 2) {
+    return out.defined() ? at::mm_out(result, tensors[0], tensors[1])
+                         : at::mm(tensors[0], tensors[1]).view(out_shape);
+  }
+
+  // Why the separate implementation for 3 matrices?
+  // The logic for three matrices is much faster when done directly
+  // Requires 1 comparison to 4 comparisons and lesser arithmetic operations
+  if (tensors.size() == 3) {
+    const auto a = tensors[0].size(0);
+    const auto b = tensors[1].size(0);
+    const auto c = tensors[2].size(0);
+    const auto d = tensors[2].size(1);
+
+    // The matrices are of size (a x b), (b x c), (c x d)
+    // cost_1 is the cost of parenthesizing (a x b) and (b x c) and then
+    // combining (c x d) cost_2 is the cost of parenthesizing (b x c) and (c x
+    // d) and then combining (a x b)
+    const auto cost_1 = (a * c) * (b + d);
+    const auto cost_2 = (b * d) * (a + c);
+
+    if (cost_1 > cost_2) {
+      return out.defined()
+          ? at::mm_out(result, tensors[0], at::mm(tensors[1], tensors[2]))
+          : at::mm(tensors[0], at::mm(tensors[1], tensors[2])).view(out_shape);
+    } else {
+      return out.defined()
+          ? at::mm_out(result, at::mm(tensors[0], tensors[1]), tensors[2])
+          : at::mm(at::mm(tensors[0], tensors[1]), tensors[2]).view(out_shape);
+    }
+  }
+
+  const auto order = matrix_chain_order(tensors);
+  const int64_t i = 0;
+  const int64_t j = n - 1;
+
+  if (out.defined()) {
+    return at::mm_out(
+        result,
+        matrix_chain_multiplication(tensors, order, i, order[i][j]),
+        matrix_chain_multiplication(tensors, order, order[i][j] + 1, j));
+  }
+  return matrix_chain_multiplication(tensors, order, i, j).view(out_shape);
+}
+
+} // namespace
+
+Tensor linalg_multi_dot(TensorList tensors) {
+  Tensor undefined;
+  return multi_dot_impl(tensors, undefined);
+}
+
+Tensor& linalg_multi_dot_out(TensorList tensors, Tensor& result) {
+  multi_dot_impl(tensors, result);
+  return result;
+}
+
+Tensor chain_matmul(TensorList matrices) {
```

Comment:
Add a comment here that this is expected to be come an alias for multi_dot, which will (slightly) change its behavior (in a way that doesn't affect backwards compatibility)